From 5fcd762831412dc0b3078c3228b98d08f2602142 Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:22:44 -0700
Subject: [PATCH] cavium: ptrace and user mem

Tweak ptrace functionality so that it can sanely deal with
the CONFIG_CAVIUM_OCTEON_USER_IO & CONFIG_CAVIUM_OCTEON_USER_MEM

Signed-off-by: Tomaso Paoletti <tpaoletti@caviumnetworks.com>
Signed-off-by: Paul Gortmaker <Paul.Gortmaker@windriver.com>
---
 arch/mips/kernel/ptrace.c |   42 +++++++++++++++++++++++++++++++++++++++++-
 1 files changed, 41 insertions(+), 1 deletions(-)

diff --git a/arch/mips/kernel/ptrace.c b/arch/mips/kernel/ptrace.c
index ec97fa4..6055e7c 100644
--- a/arch/mips/kernel/ptrace.c
+++ b/arch/mips/kernel/ptrace.c
@@ -39,6 +39,11 @@
 #include <asm/bootinfo.h>
 #include <asm/reg.h>
 
+#if defined(CONFIG_CAVIUM_OCTEON_USER_MEM) && defined(CONFIG_64BIT)
+#include "cvmx-app-init.h"
+extern cvmx_bootinfo_t *octeon_bootinfo;
+#endif
+
 /*
  * Called by kernel/ptrace.c when detaching..
  *
@@ -174,9 +179,44 @@ long arch_ptrace(struct task_struct *child, long request, long addr, long data)
 	switch (request) {
 	/* when I and D space are separate, these will need to be fixed. */
 	case PTRACE_PEEKTEXT: /* read word at location addr. */
-	case PTRACE_PEEKDATA:
+	case PTRACE_PEEKDATA: {
+#if defined(CONFIG_CAVIUM_OCTEON_USER_IO) && defined(CONFIG_64BIT)
+		unsigned long tmp;
+
+		/* check whether its a XKPHYS IO addr (we only allow the 0x80xx.. alias) */
+		if (((unsigned long)addr >> 48) == 0x8001) {
+			ret = put_user(*(unsigned long*)addr, (unsigned long __user *) data);
+			break;
+		}
+#if defined(CONFIG_CAVIUM_OCTEON_USER_MEM)
+		/* check whether its a XKPHYS MEM addr */
+		if (((unsigned long)addr >> 48) == 0x8000) {
+			ret = -EIO;
+			/* ensure that task is 64 bit */
+			if (test_thread_flag(TIF_32BIT_ADDR))
+				break;
+
+			/* extract phy addr from XKPHYS alias */
+			tmp = (unsigned long)addr - 0x8000000000000000ull;
+
+			/* check for boot-bus addr range */
+			if ((tmp >= 0x10000000) && (tmp < 0x20000000))
+				break;
+			/* this is for the dram_size comparison below */
+			if ((tmp >= 0x410000000ull) && (tmp < 0x420000000ull))
+				tmp -= 0x400000000ull;
+
+			/* verify that "addr" is within installed dram */
+			if (tmp <= ((octeon_bootinfo->dram_size<<20) - sizeof(tmp)))  {
+				ret = put_user(*(unsigned long*)addr, (unsigned long __user *) data);
+			}
+			break;
+		}
+#endif /*  defined(CONFIG_CAVIUM_OCTEON_USER_MEM) */
+#endif /* defined(CONFIG_CAVIUM_OCTEON_USER_IO) && defined(CONFIG_64BIT) */
 		ret = generic_ptrace_peekdata(child, addr, data);
 		break;
+	}
 
 	/* Read the word at location addr in the USER area. */
 	case PTRACE_PEEKUSR: {
-- 
1.5.5.1

